src/chips/merkle_sum_tree.rs

```diff
@@ -80,10 +80,11 @@ impl <F: Field> MerkleSumTreeChip<F> {
             let r2 = meta.query_advice(col_d, Rotation::next());
 
             vec![
-                s.clone() * (e.clone() * Expression::Constant(F::from(2)) * (c.clone() - a.clone())
-                    - (l1 - a)
-                    - (c - r1)),
-                s * (e * Expression::Constant(F::from(2)) * (d.clone() - b.clone())
+                s.clone()
+                    * (e.clone() * Expression::Constant(Fp::from(2)) * (c.clone() - a.clone())
```

Comment:
Yes, you are right. Here's the explanation of that:

### case no swap to be performed (current constraint)

| a   | b   | c   | d   | e   |
| --- | --- | --- | --- |:--- |
|  5  |  6  |  7  |  8  | 0    |

| l1  | l2  | r1  | r2  | 
| --- | --- | --- | --- |
|  5  |  6  |  7  |  8  |

- constraint 1 => 
``` 
e * (2 * c - a) - (l1 -a) - (c - r1) = 0
0 * (2 * 2) - 0 - 0  = 0 (satisfied!)
```
- constraint 2 => 
``` 
e * (2 * d - b) - (l2 - b) - (d - r2) = 0
0 * (2 * 2) - 0 - 0  = 0 (satisfied!)
```

### case swap to be performed (current constraint)

| a   | b   | c   | d   | e   |
| --- | --- | --- | --- |:--- |
|  5  |  6  |  7  |  8  | 1    |

| l1  | l2  | r1  | r2  | 
| --- | --- | --- | --- |
|  7  |  8  |  5  |  6  |

- constraint 1 => 
``` 
e * (2 * c - a) - (l1 -a) - (c - r1) = 0
1 * (2 * 2) - 2 - 2  = 0 (satisfied!)
```
- constraint 2 => 
``` 
e * (2 * d - b) - (l2 - b) - (d - r2) = 0
1 * (2 * 2) - 2 - 2  = 0 (satisfied!)
```
### case swap to be performed with modified constraint

| a   | b   | c   | d   | e   |
| --- | --- | --- | --- |:--- |
|  5  |  6  |  7  |  8  | 1    |

| l1  | l2  | r1  | r2  | 
| --- | --- | --- | --- |
|  7  |  8  |  5  |  6  |

- constraint 1 => 
``` 
e * ((l1 -a) - (c - r1)) = 0
1 * ((7 - 5) - (7 - 5))  = 0 (satisfied!)
```
- constraint 2 => 
``` 
e * ((l2 - b) - (d - r2)) = 0
1 * ((8 - 6) - (8 - 6))  = 0 (satisfied!)
```

When testing the constraint using the solution that you suggested, I noticed that the [`non_binary_index` test](https://github.com/summa-dev/summa-solvency/blob/mst-chip-refactor/src/circuits/tests.rs#L337) would fail. In particular, with the current version the test would generate a few errors including an error in the `swap constraint` while with the version that you suggested this error doesn't appear. That's because if we choose an `e` (swap index) that is not binary for example `e=2` the constraint won't fail

- constraint 1 => 
``` 
e * ((l1 -a) - (c - r1)) = 0
2 * ((7 - 5) - (7 - 5))  = 0 (satisfied!)
```
- constraint 2 => 
``` 
e * ((l2 - b) - (d - r2)) = 0
2 * ((8 - 6) - (8 - 6))  = 0 (satisfied!)
```

But this is ok. Since that there's already another constraint [(`bool_constraint`)](https://github.com/summa-dev/summa-solvency/blob/mst-chip-refactor/src/chips/merkle_sum_tree.rs#L62) that performs this check. 

To conclude, I'd say that the previous implementation of the `swap constraint` was "over constraining" the circuit.